sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala

```diff
@@ -277,4 +279,12 @@ class UserDefinedTypeSuite extends QueryTest with SharedSparkSession with Parque
     val udt = new TestUDT.MyDenseVectorUDT()
     assert(!Cast.canUpCast(udt, StringType))
   }
+
+  test("SPARK-29961 typeof user defined type") {
+    val schema = new StructType().add("a", new TestUDT.MyDenseVectorUDT())
+    val data = Arrays.asList(
+      RowFactory.create(new TestUDT.MyDenseVector(Array(1.0, 3.0, 5.0, 7.0, 9.0))))
+    checkAnswer(spark.createDataFrame(data, schema).selectExpr("typeof(a)"),
+      Seq(Row("array<double>")))
```

Comment:
I thought we don't need this based on the @cloud-fan suggestion: https://github.com/apache/spark/pull/26599#issuecomment-555855696 ?
In that case, we need to throw an analysis exception for unsupported types?

Reply:
Spark contributors also can benefit from this function. I suggest we make it support for `udt`s . WDYT? @clould-fan

Reply:
If we do want to support it, I think we should output `MyDenseVectorUDT` instead of `array<double>`

Reply:
this is base on how we overwrite the `sqlType` or `catalogString` of UDTs, it is `sqlType.simpleString` by default